zenoh/src/net/routing/hat/p2p_peer/token.rs

```diff
@@ -0,0 +1,482 @@
+            let mut face = &mut client_tokens[0];
+            if face.whatami != WhatAmI::Client {
+                if let Some(id) = face_hat_mut!(face).local_tokens.remove(res) {
+                    face.primitives.send_declare(RoutingContext::with_expr(
+                        Declare {
+                            interest_id: None,
+                            ext_qos: ext::QoSType::DECLARE,
+                            ext_tstamp: None,
+                            ext_nodeid: ext::NodeIdType::DEFAULT,
+                            body: DeclareBody::UndeclareToken(UndeclareToken {
+                                id,
+                                ext_wire_expr: WireExprType::null(),
+                            }),
+                        },
+                        res.expr(),
+                    ));
+                }
+                for res in face_hat!(face)
+                    .local_tokens
+                    .keys()
+                    .cloned()
+                    .collect::<Vec<Arc<Resource>>>()
+                {
+                    if !res.context().matches.iter().any(|m| {
+                        m.upgrade()
+                            .is_some_and(|m| m.context.is_some() && remote_client_tokens(&m, face))
+                    }) {
+                        if let Some(id) = face_hat_mut!(&mut face).local_tokens.remove(&res) {
+                            face.primitives.send_declare(RoutingContext::with_expr(
+                                Declare {
+                                    interest_id: None,
+                                    ext_qos: ext::QoSType::DECLARE,
+                                    ext_tstamp: None,
+                                    ext_nodeid: ext::NodeIdType::DEFAULT,
+                                    body: DeclareBody::UndeclareToken(UndeclareToken {
+                                        id,
+                                        ext_wire_expr: WireExprType::null(),
+                                    }),
+                                },
+                                res.expr(),
+                            ));
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+
+fn forget_client_token(
+    tables: &mut Tables,
+    face: &mut Arc<FaceState>,
+    id: TokenId,
+    res: Option<Arc<Resource>>,
+) -> Option<Arc<Resource>> {
+    if let Some(mut res) = face_hat_mut!(face).remote_tokens.remove(&id) {
+        undeclare_client_token(tables, face, &mut res);
+        Some(res)
+    } else if let Some(mut res) = res {
+        undeclare_client_token(tables, face, &mut res);
+        Some(res)
+    } else {
+        None
+    }
+}
+
+pub(super) fn token_new_face(tables: &mut Tables, face: &mut Arc<FaceState>) {
+    if face.whatami != WhatAmI::Client {
+        for src_face in tables
+            .faces
+            .values()
+            .cloned()
+            .collect::<Vec<Arc<FaceState>>>()
```

Comment:
Instead of cloning multiple times `src_face`, I believe you can `drain()` this iterator.

Reply:
`drain()`would not help here as the `clone()`here is mainly to avoid the double borrow. But I can indeed improve by cloning once accessing the remote tokens instead of one clone fo each token.